### PR TITLE
Fix #1071: support enums

### LINKED ISSUE #1071: Enum support
I had already asked this in the mailing list, but it seems to be lost there.  I had tried to use enums for some algorithms using lots of bit manipulation. For that I have to manage many integer constants, which I had tried to efficiently group with some well readable syntax like that:

```python
class SomeConstantsContainer:
    SOME_NAME = 0x1
    SOME_OTHER_CONSTANT = 0x2
    AND_ANOTHER_CONSTANT = 0x4
```
or that:

```python
@enum.unique
class SomeConstantsContainer(enum.IntEnum):
    SOME_NAME = 0x1
    SOME_OTHER_CONSTANT = 0x2
    AND_ANOTHER_CONSTANT = 0x4
```

Unfortunately both fails in nopython-mode. Both, the container and the value are only recognized as python objects. I suppose I would have to write something like this, to get full compiled speed:

```
SOME_STUPID_CONSTANT = np.int64(0x1)
ANOTHER_STUPID_CONSTANT = np.int64(0x2)
```

For usability and readability I don't like that. I have different constants with different meanings, which I'd like to group visibly together in any form. I'm also a big fan, when the IDE can help me autocompleting the names of the constants, by picking them from the enums. 

Right now, there seems to be now way to efficiently handle the grouping. Only one suggestion on stackoverflow, using closures, came somewhat close: http://stackoverflow.com/questions/28634652/best-way-to-handle-multiple-constants-in-a-container-with-numba So, support for enums would really be awesome!

### PR DESCRIPTION
**Pull request by pitrou**
This adds support for `Enum` and `IntEnum` classes.

### CODE REVIEW

File: numba/tests/enum_usecases.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+class Planet(Enum):
+    MERCURY = (3.303e+23, 2.4397e6)
+    VENUS   = (4.869e+24, 6.0518e6)
+    EARTH   = (5.976e+24, 6.37814e6)
+    MARS    = (6.421e+23, 3.3972e6)
+    JUPITER = (1.9e+27,   7.1492e7)
+    SATURN  = (5.688e+26, 6.0268e7)
+    URANUS  = (8.686e+25, 2.5559e7)
+    NEPTUNE = (1.024e+26, 2.4746e7)
```
Comment (seibert) [inline numba/tests/enum_usecases.py]: I see that `PLUTO` has been dutifully excluded as per the IAU.  :)

**Comment by pitrou** [comment]
Note the Python 2.7 tests are currently failing because of this issue:
https://bitbucket.org/ambv/singledispatch/issues/8/inconsistent-hierarchy-with-enum

File: numba/tests/test_typeof.py [hunk 2]
```diff
@@ -213,6 +215,34 @@ def test_namedtuple(self):
...
+        self.assertEqual(tp_red, typeof(Color.blue))
+        tp_choc = typeof(Shake.chocolate)
+        self.assertEqual(tp_choc, types.EnumMember(Shake, types.intp))
+        self.assertEqual(tp_choc, typeof(Shake.mint))
+        self.assertNotEqual(tp_choc, tp_red)
+
+        with self.assertRaises(ValueError) as raises:
+            typeof(HeterogenousEnum.red)
+        self.assertEqual(str(raises.exception),
+                         "Cannot type heterogenous enum: got value types complex128, int64")
```
Comment (pitrou) [inline numba/tests/test_typeof.py]: Note: I should fix this to allow int32 as well.

**Comment by muxator** [comment]
Hi, sorry for commenting on this old issue, but I am having problems with the documentation section of this merged PR.

It is really hard to understand what "both `enum.Enum` and `enum.IntEnum` subclasses are supported" without any example or reference in the documentation.

The intuitive way, of `@jitclass`-ing a Python Enum throws an `AttributeError` in my code base. Apparently I am not the only one, see [this stack overflow question](https://stackoverflow.com/questions/61155860/can-i-compile-a-python-enum-into-a-numba-jitclass) for a more complete example.

Are there any limitations/preconditions to be met? This should really go in the user documentation.

**Comment by esc** [comment]
@muxator thanks for raising this issue! Since this is an old PR here, I would recommend getting a discussion going on https://numba.discourse.group/ to figure out what works and what doesn't and then submitting a fresh PR with any necessary modifications to the Numba documentation.

**Comment by Jasha10** [comment]
For example, this works:
```
import numba
from enum import Enum

class MyEnum(Enum):
    Red = 123
    Blue = 456

@numba.njit
def check_isblue(e):
    assert e == MyEnum.Blue

check_isblue(MyEnum.Blue)
```

See the file [`numba/tests/test_enums.py`](https://github.com/numba/numba/blob/master/numba/tests/test_enums.py) introduced by this pull request.

**Pull request merged**